src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -567,4 +571,64 @@ public virtual void PrintMessageNormalized(TextWriter writer, string message, bo
         else if (!removeLastNewLine)
             writer.WriteLine();
     }
+
+    /// <summary>
+    /// Adds a new runtime option. Throws if option already exists.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="name"></param>
+    /// <param name="value"></param>
+    public void AddOption<T>(string name, T value)
+    {
+        Options.Add(name, value);
+    }
+
+    /// <summary>
+    /// Trys to add a new runtime option.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="name"></param>
+    /// <param name="value"></param>
+    /// <returns>True if successful. False otherwise.</returns>
+    public bool TryAddOption<T>(string name, T value)
+    {
+        if (Options.ContainsKey(name))
+            return false;
+        AddOption<T>(name, value);
+        return true;
+    }
+
+    public void AddOrOverwriteOption<T>(string name, T value)
+    {
+        if (!Options.ContainsKey(name))
+            AddOption<T>(name, value);
+        else
+            Options[name] = value;
+    }
+
+    public T GetOption<T>(string name)
+    {
+        return (T)Options[name];
```

Comment:
>return (T)Options[name]; [](http://example.com/codeflow?start=8&length=24)

This can throw invalid cast if cannot convert object to T. Is it ok? Or better to make it TryGetOptionValue instead and return true or false depending on the existence of the option and valid cast?

Reply:
I think its ok to throw the invalid cast, but what do you think?

I am also happy to add a `TryGet...` as well, but I don't think we should just remove this one. Thoughts?

Also, since I have the `GetOptionOrDefault` should I use that instead of the plain `Get` and change that to `Try`?

Reply:
> since I have the GetOptionOrDefault should I use that instead of the plain Get and change that to Try?

I think this will be better.

Reply:
Fixed.